Breaking changes policy

plavjanik opened the pull request:
Hello,

this a draft of the proposal for the breaking changes policy based on the discussion with TSC and on the architecture meeting earlier (see https://github.com/zowe/community/issues/1466).

Please provide your suggestion in this PR. Another discussion will be held at TSC next week.

It is inspired by Node.js - https://github.com/nodejs/node/blob/master/doc/contributing/collaborator-guide.md#breaking-changes and adapted to the needs of Zowe.

balhar-jakub commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+![Schedule for the nearest releases](/Technical-Steering-Committee/major_releases.png)
+
+At least two TSC members must approve backward-incompatible changes to the under development branch or current active or maintenance release. Changes to the development branch will be breaking changes in the upcoming active release. Breaking changes to active or maintenance release should be avoided unless necessary (for example critical fixes or security vulnerabilities).
```
I would be careful about the under development branch. It's possible that some squads develop in under development branch including breaking changes and cherry-pick to the active and major branches.

balhar-jakub commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+Apply the `notable change` label to all pull requests that introduce Documentation-Only Deprecations. Such deprecations have no impact on code execution. Thus, they are not breaking changes.
+
+Documentation-Only Deprecations can land in a minor release. They can change to a Runtime Deprecation after six months. They can not change to a End-of-Life until the next major release and need to be in End-of-Life at least for another six months.
```
I may be a bit confused by the last part. 

With respect to the next major release, it's ok to change to the EOL and that means it's not available in the next major release. 

In the current major release it needs to be supported until the end of the maintenance phase.

balhar-jakub commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+Documentation-Only Deprecations can land in a minor release. They can change to a Runtime Deprecation after six months. They can not change to a End-of-Life until the next major release and need to be in End-of-Life at least for another six months.
+
+No API can change to End-of-Life without going through a Runtime Deprecation cycle. There is no rule that deprecated code must progress to End-of-Life. Documentation-Only and Runtime Deprecations can remain in place for an unlimited duration.
```
Valid only for projects reaching the LTS maturity. For those that achieved lesser maturity the full runtime deprecation cycle can't be required. We may need to look for simplified version.

plavjanik replied:
Good comment. I added a sentence about that this policy applies to the projects that are not in incubation. We can discuss if we want to have a policy for incubation projects.

balhar-jakub commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+1. A planned breaking change is discussed at Zowe architecture call and Zowe playback demo event.
+
+2. The breaking change is brought to TSC and approved.
```
I don't think that TSC should approve all breaking changes. This is the responsibility of the squads as long as the change doesn't get to the customers before next major release.

plavjanik replied:
Agree like with the previous comment about TSC decisions. I will reword it.

plavjanik replied:
Change has been done.

balhar-jakub commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+Each squad documents for the components owned by the squads in [Zowe Docs](https://docs.zowe.org/):
+
+ - What is public and internal API
```
You mean definition of how the squad thinks about what is public and internal?

plavjanik replied:
Yes. I tried to reword it better.

balhar-jakub reviewed (commented):
Thanks for the preparation on the policy. I believe there are still things to discuss and agree on, but overall it's definitely going in the right direction.

plavjanik commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+Breaking changes are sometimes necessary to keep the Zowe project in a good shape. This document provides guidelines and policy about what is breaking change, how it can be introduced so the change will have minimal disruptive effects on the Zowe users and extenders.
+
+This policy will be affective with the Zowe V2 release. It will affect how changes to V2 are introduced. Breaking changes to V2 APIs and behavior can be made in V3 at earliest.
```
The first sentence is here to emphasize this approach and point out "minimal disruptive effects" since that has not been always the case.

plavjanik commented on Technical-Steering-Committee/guidelines/breaking-changes.md:
```diff
@@ -0,0 +1,136 @@
+
+Exceptions are undocumented behaviors that are relied on by the community (e.g. `zowe jobs submit` command returns the job ID without other data or messages). Such occurrences should be reported via issues and documented over time.
+
+Sometimes, there is disagreement about whether functionality is internal or public. In those cases, the TSC makes a determination.
```
I agree. I will reword it. This is coming from the Node.js document. Zowe is rather set of loosely coupled components than a platform like Node.js and it make sense for each squad to have ownership. The appeal to TSC will make possible that changes that might be troublesome from the point of view the other squads can be discussed appropriately.

plavjanik replied:
Change has been done.

plavjanik commented:
Additional changes:
- Added a requirement of deprecation identifier for referencing the deprecation
- Added note about the compatibility of client libraries

plavjanik commented:
Thank you, @1000TurquoisePogs, for mentioning the configuration changes. It is very important to call it out explicitly. I have reused most of your words. I have added one point about migrating configuration from an older version of schema to a new one. It will be sometimes necessary to make bigger changes to the schema and the approach that I have seen in several projects is automatic migration from the old format to the new one. So, we can have a new format that can do even bigger structural changes under the condition that the old version of the format can be used without any change and migrated to the new version.

balhar-jakub commented:
The changes linked to the Breaking policy were moved to another PR.
